substrate/client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -172,8 +240,224 @@ fn addresses_to_peer_ids(addresses: &HashSet<Multiaddr>) -> HashSet<PeerId> {
+					})
+					.collect::<Result<HashSet<Multiaddr>, Self::Error>>()?;
+				Ok((authority_id, addresses))
+			})
+			.collect::<Result<HashMap<AuthorityId, HashSet<Multiaddr>>, Self::Error>>()?;
+
+		let peer_id_to_authority_ids = value
+			.peer_id_to_authority_ids
+			.into_iter()
+			.map(|(peer_id, authority_ids)| {
+				let peer_id = PeerId::try_from(peer_id)?;
+				Ok((peer_id, authority_ids.into_iter().collect::<HashSet<AuthorityId>>()))
+			})
+			.collect::<Result<HashMap<PeerId, HashSet<AuthorityId>>, Self::Error>>()?;
+
+		Ok(AddrCache { authority_id_to_addresses, peer_id_to_authority_ids, on_change: None })
+	}
+}
+
+/// A (de)serializable version of [`PeerId`] that can be used for serialization,
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Hash)]
+#[serde(transparent)]
+struct SerializablePeerId {
+	#[serde_as(as = "serde_with::hex::Hex")]
+	bytes: Vec<u8>,
+}
+
+impl From<PeerId> for SerializablePeerId {
+	fn from(peer_id: PeerId) -> Self {
+		Self { bytes: peer_id.to_bytes() }
+	}
+}
+impl TryFrom<SerializablePeerId> for PeerId {
+	type Error = Error;
+
+	fn try_from(value: SerializablePeerId) -> Result<Self, Self::Error> {
+		PeerId::from_bytes(&value.bytes)
+			.map_err(|e| Error::EncodingDecodingAddrCache(e.to_string()))
+	}
+}
+
+/// A (de)serializable version of [`Multiaddr`] that can be used for serialization,
+#[serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(transparent)]
+struct SerializableMultiaddr {
+	/// `Multiaddr` holds a single `LiteP2pMultiaddr`, which holds `Arc<Vec<u8>>`.
+	#[serde_as(as = "serde_with::hex::Hex")]
+	bytes: Vec<u8>,
+}
+impl From<Multiaddr> for SerializableMultiaddr {
+	fn from(multiaddr: Multiaddr) -> Self {
+		Self { bytes: multiaddr.to_vec() }
+	}
+}
+impl TryFrom<SerializableMultiaddr> for Multiaddr {
+	type Error = ParseError;
+
+	fn try_from(value: SerializableMultiaddr) -> Result<Self, Self::Error> {
+		Self::try_from(value.bytes)
+	}
+}
+
+/// Writes the content to a file at the specified path.
+fn write_to_file(path: &PathBuf, content: &str) -> io::Result<()> {
+	let mut file = File::create(path)?;
+	file.write_all(content.as_bytes())?;
+	file.flush()
+}
+
+/// Reads content from a file at the specified path and tries to JSON deserializes
+/// it into the specified type.
+fn load_from_file<T: DeserializeOwned>(path: &PathBuf) -> io::Result<T> {
+	let file = File::open(path)?;
+	let reader = BufReader::new(file);
+
+	serde_json::from_reader(reader).map_err(|e| {
+		error!(target: super::LOG_TARGET, "Failed to load from file: {}", e);
+		io::Error::new(io::ErrorKind::InvalidData, e)
+	})
+}
+
+/// Asynchronously writes content to a file using a background thread.
+/// Multiple consecutive writes in quick succession (before the current write is completed)
+/// will be **throttled**, and only the last write will be performed.
+#[derive(Clone)]
+pub struct ThrottlingAsyncFileWriter {
```

Comment:
This should probably be moved elsewhere! At the very least to its own file! But maybe to another "common"/"foundation" crate? Which one?

Reply:
Agree that file operations ideally should be outside of `authority-discovery` crate. But honestly don't know where is the "proper" place.